Yet More Tool Execution Optimizations

jmchilton opened the pull request:
See individual commits for details.

Builds on the declined #1197 (with fixes from @nsoranzo) with many new optimizations large and small.

jmchilton committed:
Tool optimization - don't set state to queued twice.

Just keep the dataset in the correct NEW state until it has actually been queued. Addresses FIXME comment added by James 7 years ago in https://github.com/galaxyproject/galaxy/commit/4c3db1af95fb0520960046ae549462aebd78b326.

This behavior feels correct to me, but it does have ramifications in the GUI. I had previously never actually seen a dataset in the "NEW" state.

Saves an extra flush per dataset, on sqlite this translates to 50ms per dataset for me.

jmchilton committed:
Tool optimization and comment improvement.

jmchilton committed:
Tool optimization - remove another seemingly needless flush.

Saves around a teenth of a second per job.

Before:

 - galaxy.tools.execute DEBUG 2015-12-01 13:00:56,984 Tool [cat1] created job [24] (435.468 ms)
 - galaxy.tools.execute DEBUG 2015-12-01 13:01:17,296 Tool [cat1] created job [25] (341.687 ms)
 - galaxy.tools.execute DEBUG 2015-12-01 13:01:32,896 Tool [cat1] created job [26] (354.138 ms)

After -

 - galaxy.tools.execute DEBUG 2015-12-01 12:58:51,154 Tool [cat1] created job [21] (261.932 ms)
 - galaxy.tools.execute DEBUG 2015-12-01 12:59:11,354 Tool [cat1] created job [22] (278.699 ms)
 - galaxy.tools.execute DEBUG 2015-12-01 12:59:43,105 Tool [cat1] created job [23] (296.438 ms)

jmchilton committed:
Micro-optimizations for adding many datasets to history...

... during tool execution. Build method once instead of in each function call, use imap instead of map since we don't need a list, remove some duplicated checks. Frankly this is all stuff Python is probably doing anyway - but in case it doesn't and just so the eye doesn't jump to these optimizations again.

Timings before and after for a section of tool action execute that includes this additon show that this might have a small effect.

Before:

 - galaxy.tools.actions INFO 2015-12-01 13:11:39,446 Add outputs to history (127.379 ms)
 - galaxy.tools.actions INFO 2015-12-01 13:12:06,598 Add outputs to history (137.029 ms)
 - galaxy.tools.actions INFO 2015-12-01 13:12:23,931 Add outputs to history (118.489 ms)

After:

 - galaxy.tools.actions INFO 2015-12-01 13:13:09,999 Add outputs to history (99.456 ms)
 - galaxy.tools.actions INFO 2015-12-01 13:13:38,573 Add outputs to history (126.131 ms)
 - galaxy.tools.actions INFO 2015-12-01 13:13:54,538 Add outputs to history (137.643 ms)
 - galaxy.tools.actions INFO 2015-12-01 13:14:12,516 Add outputs to history (101.451 ms)

jmchilton committed:
Tool optimization comments for future reference.

Including replacing a log statement with a comment.

jmchilton committed:
Tool execution optimization - remove another unneeded flush.

Flush the job and its output datasets all at the same time in one transaction.

Inuitively it makes sense that this should work and the timings seem to improve, triming another quater or third of the submission time per job.

Before:

 - galaxy.tools.execute DEBUG 2015-12-01 12:58:51,154 Tool [cat1] created job [21] (261.932 ms)
 - galaxy.tools.execute DEBUG 2015-12-01 12:59:11,354 Tool [cat1] created job [22] (278.699 ms)
 - galaxy.tools.execute DEBUG 2015-12-01 12:59:43,105 Tool [cat1] created job [23] (296.438 ms)

After:

 - galaxy.tools.execute DEBUG 2015-12-01 13:22:00,649 Tool [cat1] created job [39] (265.547 ms)
 - galaxy.tools.execute DEBUG 2015-12-01 13:22:27,720 Tool [cat1] created job [40] (214.905 ms)
 - galaxy.tools.execute DEBUG 2015-12-01 13:21:40,225 Tool [cat1] created job [38] (198.936 ms)
 - galaxy.tools.execute DEBUG 2015-12-01 13:22:44,076 Tool [cat1] created job [41] (213.096 ms)

jmchilton committed:
Micro optimization in tool execute - eliminate repeated trans.app fetch.

jmchilton committed:
Instrument upload action for timing.

 - Checked and cleanup datasets ~500ms
 - Create job ~200ms
 - Persisting ~20 ms

jmchilton committed:
Tool optimization - eliminate repeated role generation.

For cat1 with 5 inputs - this shaves about 100 ms off of job execution time.

Before:

 galaxy.tools.execute DEBUG 2015-12-01 15:08:56,745 Tool [cat1] created job [171] (545.776 ms)
 galaxy.tools.execute DEBUG 2015-12-01 15:08:31,871 Tool [cat1] created job [170] (478.813 ms)

After :

 galaxy.tools.execute DEBUG 2015-12-01 15:07:11,401 Tool [cat1] created job [168] (371.722 ms)
 galaxy.tools.execute DEBUG 2015-12-01 15:07:42,060 Tool [cat1] created job [169] (333.296 ms)

jmchilton committed:
Fix location of timer generation for fairer comparison in next commit.

jmchilton committed:
Further cache user roles across batched tool executions.

martenson commented on config/galaxy.ini.sample:
```diff
@@ -832,6 +832,20 @@ use_interactive = True
+# web processes, not job handlers.
+#tool_submission_burst_threads = 1
+
+# If the above setting is to set to greater than 1, that number of 
```
You can reword the first sentence as 'If the tool_submission_burst_threads setting above is greater than 0'

martenson commented on config/galaxy.ini.sample:
```diff
@@ -832,6 +832,20 @@ use_interactive = True
+#tool_submission_burst_threads = 1
+
+# If the above setting is to set to greater than 1, that number of 
+# threads will be used to process requests resulting in greater than
```
The rest of the paragraph is cryptic to me.

martenson commented:
looks great! :+1:

jmchilton committed:
Fix collection output for less flushing in tool action handle_output.

jmchilton commented:
So the problem was that current datasets do need to be flushed and have an ID before calling the object store create code (sometimes, it actually works in many cases including my manual testing when collecting the timings).

So I have rebased and replaced 619f334 with a1cc98076df21bdca8ee157010ad3dec888ed733 and added a comment about this problem. Unfortunately, this renders the commit message and timings in many subsequent commits incorrect - still I guess they still all improve the situation and shouldn't be broken by an extra flush.

This has also been rebased on top of #1212, so when that is merged the list of commits here will simply shrink to something more manageable.

jmchilton commented:
Also looks like I need to fix the unit tests.

bgruening commented:
@jmchilton sorry we also need a rebase here. This looks really awesome!

jmchilton commented:
Rebased.

bgruening commented:
:+1: from me as well!
